colossalai/zero/sharded_optim/sharded_optim_v2.py

```diff
@@ -52,20 +56,24 @@ def __init__(self,
                                              growth_interval=growth_interval,
                                              hysteresis=hysteresis,
                                              max_scale=max_scale)
-        self._found_overflow: Tensor = torch.FloatTensor([0]).to(self.device)
+        self._found_overflow: Tensor = torch.FloatTensor([0]).to(torch.cuda.current_device())
 
-        # Store fp32 params
+        # Store fp32 param shards
         self.master_params: Dict[Parameter, Tensor] = {}
 
         for group in optimizer.param_groups:
             for p in group['params']:
-                if hasattr(p, 'ca_attr'):
-                    assert p.ca_attr.is_sharded, 'ShardedAdam can be only used with sharded model'
-                    self.master_params[p] = p.ca_attr.payload(self.device)
-                else:
-                    self.master_params[p] = p.data.to(device=self.device)
-                if torch.is_floating_point(self.master_params[p]) and self.master_params[p].dtype != torch.float:
-                    self.master_params[p] = self.master_params[p].to(torch.float)
+                assert hasattr(p, 'col_attr'), 'The parameter must be wrapped with ShardedParam'
+                is_param_sharded = p.col_attr.data.is_sharded
+                if not is_param_sharded:
+                    # Param is no sharded
+                    # As we only store param shard, we shard it here
+                    self.shard_strategy.shard([p.col_attr.data])
+                self.master_params[p] = cast_tensor_to_fp32(p.col_attr.data.payload).to(self.device)
+                if not is_param_sharded:
+                    # In this branch, there's no need to shard param
+                    # So we gather here
+                    self.shard_strategy.gather([p.col_attr.data])
 
     def step(self, *args, **kwargs):
         # unscale grads if scaled
```

Comment:
Where did you shard the momentum and variance data?

Reply:
As p.data and p.grad are sharded, optimizer states are sharded naturally.

Reply:
Pytorch optimizers do a lazy init of optimizer states, whose shape depends on the shape of p.data when doing optimizer.step().

Reply:
I see